pxtsim/simlib.ts

```diff
@@ -37,34 +43,56 @@ namespace pxsim {
         constructor(private runtime: Runtime, private valueToArgs?: EventValueToActionArgs<T>) { }
 
         private start(id: number | string, evid: number | string, create: boolean) {
-            let k = id + ":" + evid;
+            let k = (this.backgroundHandlerFlag ? "back:" : "") + id + ":" + evid;
             let queue = this.queues[k];
             if (!queue) queue = this.queues[k] = new EventQueue<T>(this.runtime, this.valueToArgs);
             return queue;
         }
 
         listen(id: number | string, evid: number | string, handler: RefAction) {
             let q = this.start(id, evid, true);
-            q.handler = handler;
+            if (this.backgroundHandlerFlag)
+                q.addHandler(handler);
+            else
+                q.setHandler(handler);
+            this.backgroundHandlerFlag = false;
+        }
+
+        // only for background handlers
+        remove(handler: RefAction) {
+            this.backgroundHandlerFlag = false;
+            Object.keys(this.queues).forEach((k: string) => {
+                if (k.startsWith("back"))
+                    this.queues[k].removeHandler(handler);
+            });
         }
 
         queue(id: number | string, evid: number | string, value: T = null) {
             // special handling for notify one
             const notifyOne = this.notifyID && this.notifyOneID && id == this.notifyOneID;
             if (notifyOne)
                 id = this.notifyID;
-
-            // grab queue and handle
+            // first, background
+            this.backgroundHandlerFlag = true;
```

Comment:
This flag seems to be used for two different purposes - for setting handlers and for running them. Two flags would be better, especially when handlers register new handlers.